cla_backend/apps/cla_eventlog/management/commands/find_and_delete_old_cases.py

```diff
@@ -0,0 +1,41 @@
+import sys
+from django.core.management.base import BaseCommand
+from dateutil.relativedelta import relativedelta
+
+from legalaid.models import Case
+from cla_butler.tasks import DeleteOldData
+
+
+class FindAndDeleteCasesUsingCreationTime(DeleteOldData):
+    def get_eligible_cases(self):
+        two_years = self.now - relativedelta(years=2)
+
+        return Case.objects.filter(created__lte=two_years).exclude(log__created__gte=two_years)
+
+
+class Command(BaseCommand):
+    help = (
+        "Find or delete cases that are 2 years old or over that were not deleted prior to the task command being fixed"
+    )
+
+    def handle(self, *args, **kwargs):
+        instance = FindAndDeleteCasesUsingCreationTime()
+        cases = instance.get_eligible_cases()
+        django_command = sys.argv[1]
+        if args and args[0] == "delete":
+            if len(args) > 1 and args[1] == "no-input":
+                instance.run()
+            elif django_command == "test":
+                instance.run()
+            else:
+                answer = raw_input(
+                    "Number of cases that will be deleted: {0}\nAre you sure about this? (Yes/No) ".format(
+                        cases.count()
+                    )
+                )
+                if answer == "Yes":
+                    instance.run()
+        elif sys.argv[1] == "test":
+            return cases
+        else:
+            print("Number of cases to be deleted: " + str(cases.count()))
```

Comment:
```        django_command = sys.argv[1]
        if args and args[0] == "delete":
            if len(args) > 1 and args[1] == "no-input":
                instance.run()
            elif django_command == "test":
                instance.run()
            else:
                answer = raw_input(
                    "Number of cases that will be deleted: {0}\nAre you sure about this? (Yes/No) ".format(
                        cases.count()
                    )
                )
                if answer == "Yes":
                    instance.run()
        elif sys.argv[1] == "test":
            return cases
        else:
            print("Number of cases to be deleted: " + str(cases.count()))```
  
I find this quite hard to follow, is there a way to simplify by doing something like 
if is_report_operation then
  show report
if is_delete_operation then
   delete cases

If you do that then you don't need to make any exception for tests, as they could mock the `is_delete_operation` or pass in the correct arguments

Reply:
E.g. for the `is_report_operation`, you can only return a string when running `command.execute()`, so I have to make conditions depending on if it's being run in terminal or via a test i.e. we want to return queryset when run in a test and can only print the number of cases in terminal

I was thinking I could refactor in such a way where I can make the first conditions based upon the environment it's in 🤔

Reply:
Yeah I think it could maybe be more readable, but fundamentally it's complex.

The routes we have are:
- cli no args -> find cases and print the count
- cli "delete" arg -> find cases, prompt for confirmation, delete the cases
- cli "delete noinput" args -> find cases, delete the cases
- via test command, no args -> find cases and return queryset
- via test command, "delete" arg -> find cases, delete the cases

So yeah I agree with the direction @said-moj's going, but maybe not quite with some of the conclusions about the testing thing (I think maybe I'm erring on the side of avoiding mocking the actual moving part that should be under test? I'm not sure).

You could add `find()` and `find_and_delete()` methods to the management command (alongside `handle()`) and hand off to those with your `if args and args[0] == "delete":`

So it'd look something like
```
if args and args[0] == "delete":
    return self.find_and_delete()
else:
    return self.find()
```
with all the delete logic you have now inside one method, and the fallback to either return qs or print count inside the other. No less code, still special-casing tests behaviour, but potentially much easier to read.

Reply:
I've refactored it so the first condition is based on what environment it is run in. Hopefully this is more readable?